feat: Update GitHub setup flow to include incidents detection PUL-1674 (#195)
codacy/pulse-user-docs: docs/one-click-integrations/github-integration.md

    ![Pulse GitHub App installed successfully](images/github-installed-ok.png)

1.  Choose the strategy to detect **deployments** that best fits your workflows. See the [section below](#deployment-detection-strategy) for a detailed description of each option.

    ![Choosing a deployment detection strategy](images/github-strategy.png)

1.  Choose the strategy to detect **incidents** that best fits your workflows. See the [section below](#incident-detection-strategy) for a detailed description of each option.

>     !!! note

Author: @mrfyda I also added a note here in the setup flow, although this could be seen as a UI failure.

> The following is a detailed description of how the Pulse GitHub integration automatically detects deployment using each detection strategy:

Reviewer: Since there are sections with titles following this paragraph, I suggest reformatting it to introduce links.

This would also require adding IDs to the sections below and tweaking the titles to fit the list format. Just an idea; we can do this further on.

```suggestion
The Pulse GitHub integration can detect deployments using the following strategies:

-   [Merged pull requests](#)  (based on default branch)
-   [Semantic versioning tags](#)
-   [CLI or API](#) (don't detect deployments automatically)
```

Author: Updated!

-   Pulse associates all GitHub teams of the person who creates a Git tag with the corresponding deployment, excluding teams with less than two members. Pulse only takes changes to GitHub teams into account on Git tags created after those changes.

### Use the CLI or API (don't detect deployments automatically)

-   Pulse doesn't detect deployments automatically using GitHub events.

    This is useful if none of the automatic deployment detection strategies match your workflow and you must have control over the way Pulse tracks your deployments.

-   In this case, you must send to Pulse the information about your **deployments** and the corresponding **changes** using the [Pulse CLI](../cli/cli.md) or the [Ingestion API](https://ingestion.pulse.codacy.com/v1/api-docs).

> ## Automatic incident detection strategies {: id="incident-detection-strategy"}

Author: @afsmeira Can you please validate this section? 🙏 
I put it similar to Bitbucket, but I'm not sure if GitHub works exactly the same...

Reviewer 2: Rafael already commented on the details that are not completely correct. 👍

> The following is a detailed description of how the Pulse GitHub integration automatically detects incidents:

Reviewer: Same as above, we could expand this into a list of links.

Author: Updated.

### Use pull request reverts (based on default branch)

> -   Pulse bases incident detection on [pull request reverts](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/reverting-a-pull-request).

Author: @afsmeira I added this link to GitHub documentation, but it's not as detailed as Bitbucket. For example, it doesn't refer to the created branch when a pull request is reverted.

Reviewer 2: When reverting a PR, GitHub will not allow you to customize the branch name. I guess that's why they don't mention it.

> -   Pulse considers an incident any pull request that **targets the default branch** of the repository merged from a branch which name starts with `revert-pr-`, getting the number of the reverted pull request from the branch name, `revert-pr-<pull request number>`. If you change the name of the branch created by GitHub when you revert a pull request, Pulse may not be able to obtain the incident data correctly.

Reviewer: Small tweak here. Apparently it's OK to [use whose with objects](https://dictionary.cambridge.org/grammar/british-grammar/whose#:~:text=Whose%20in%20relative%20clauses), just found out 😄 

```suggestion
-   Pulse considers an incident any pull request that **targets the default branch** of the repository merged from a branch whose name starts with `revert-pr-`, getting the number of the reverted pull request from the branch name, `revert-pr-<pull request number>`. If you change the name of the branch created by GitHub when you revert a pull request, Pulse may not be able to obtain the incident data correctly.
```

Author: Nice catch! 🤓

> -   Pulse considers an incident any pull request that **targets the default branch** of the repository merged from a branch which name starts with `revert-pr-`, getting the number of the reverted pull request from the branch name, `revert-pr-<pull request number>`. If you change the name of the branch created by GitHub when you revert a pull request, Pulse may not be able to obtain the incident data correctly.

Reviewer 2: In GitHub it's not possible to customize the name of a branch that was created via the revert feature. But I guess the warning still makes sense, since users can manually follow this convention.

> -   Pulse considers an incident any pull request that **targets the default branch** of the repository merged from a branch which name starts with `revert-pr-`, getting the number of the reverted pull request from the branch name, `revert-pr-<pull request number>`. If you change the name of the branch created by GitHub when you revert a pull request, Pulse may not be able to obtain the incident data correctly.

Author: Thanks! I'll keep the warning. 👍

-   The incident creation date is the timestamp when the reverted pull request was initially merged. If Pulse can't get the reverted pull request number from the branch name, the incident creation date is the timestamp of the first commit to the incident pull request.
-   Pulse associates incidents to the system matching the repository name.

### Don't detect incidents via GitHub

> -   Pulse doesn't detect incidents automatically using GitHub events.

Reviewer: * Taken out of context, this may sound like an absolute statement. I'd qualify it with an `If...`.
  * Now that I'm re-reading the above paragraphs, we may want to qualify all of them with an initial `If`. Thoughts?
* Since this is a lone bullet point, we can consider dropping it and turning this into a paragraph

```suggestion
-   If you choose this option, Pulse doesn't detect incidents automatically using GitHub events.
```

Author: Yep, I see what you mean. If I would add an `If..` to one option, I would have to add it to all, or move it to a [stem sentence](http://dita-archive.xml.org/wiki/stem-sentences).
A stem sentence would apply here as we have sections below... I think I would leave as if for now. 🙏

-   Pulse associates all GitHub teams of the person who creates a Git tag with the corresponding deployment, excluding teams with less than two members. Pulse only takes changes to GitHub teams into account on Git tags created after those changes.

### Use the CLI or API (don't detect deployments automatically) {: id="gh-deploy-cli-api"}

-   Pulse doesn't detect deployments automatically using GitHub events.

    This is useful if none of the automatic deployment detection strategies match your workflow and you must have control over the way Pulse tracks your deployments.

-   In this case, you must send to Pulse the information about your **deployments** and the corresponding **changes** using the [Pulse CLI](../cli/cli.md) or the [Ingestion API](https://ingestion.pulse.codacy.com/v1/api-docs).

## Automatic incident detection strategies {: id="incident-detection-strategy"}

The Pulse GitHub integration detects incidents automatically using [pull request reverts](#gh-incident-pr-revert). You can also choose [not to detect incidents via GitHub](#gh-incident-not-detect).

> !!! note

Author: @mrfyda I added a note here with the identified limitation. Can you please validate?

Reviewer 3: @pedronetocodacy can you have a look?

    <td>
        <p>Deployments:</p>
        <ul>
            <li><code>deploy_id</code>: pull request identifier</li>

>             <li><code>timestamp_created</code>: pull request merged date</li>

Reviewer: See [this comment](https://github.com/codacy/pulse-user-docs/pull/195/files#r1108223461) further down

```suggestion
            <li><code>timestamp_created</code>: merge date of the pull request</li>
```

            <li><code>system</code>: repository name</li>
        </ul>
    </td>
    <td>Deployment frequency and Change failure rate on the <a href="../../metrics/accelerate/">Accelerate Overview dashboard</a></td>
</tr>
<tr>
<td>Pull requests</td>
    <td>
        <p>Incidents:</p>
        <ul>

>             <li><code>incident_id</code>: pull request number</li>

Reviewer 3: https://github.com/codacy/pulse-user-docs/pull/194/files#r1107053946

```suggestion
            <li><code>incident_id</code>: pull request identifier</li>
```

Reviewer 2: Yeah, as I mentioned in the Bitbucket PR, GitHub actually has a unique PR identifier which we use for this purpose.

>             <li><code>timestamp_created</code>: reverted pull request merged date</li>

Reviewer: I'd. break up this [noun string](https://www.plainlanguage.gov/guidelines/words/avoid-noun-strings/)

```suggestion
            <li><code>timestamp_created</code>: merge date of the reverted pull request</li>
```

>             <li><code>timestamp_resolved</code>: pull request merged date</li>

Reviewer: Same here:

```suggestion
            <li><code>timestamp_resolved</code>: merge date of the pull request</li>
```
